From 452829dcd8772e24c2b7d75e05a57eb577697996 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:08:00 -0400
Subject: [virt] kvm/ppc: define a linux pte lookup function

Message-id: <1410545655-205645-252-git-send-email-dzickus@redhat.com>
Patchwork-id: 94214
O-Subject: [RHEL7 PATCH 251/626] kvm: powerpc: define a linux pte lookup function
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit f5e3fe091f5238459752a81b478398b7cb22e575
Author: Bharat Bhushan <r65777@freescale.com>
Date:   Fri Nov 15 11:01:15 2013 +0530

    kvm: powerpc: define a linux pte lookup function

    We need to search linux "pte" to get "pte" attributes for setting TLB in KVM.
    This patch defines a lookup_linux_ptep() function which returns pte pointer.

    Signed-off-by: Bharat Bhushan <bharat.bhushan@freescale.com>
    Reviewed-by: Scott Wood <scottwood@freescale.com>
    Signed-off-by: Alexander Graf <agraf@suse.de>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/include/asm/pgtable.h b/arch/powerpc/include/asm/pgtable.h
index 78784a7..4fbc042 100644
--- a/arch/powerpc/include/asm/pgtable.h
+++ b/arch/powerpc/include/asm/pgtable.h
@@ -312,6 +312,27 @@ extern int gup_hugepte(pte_t *ptep, unsigned long sz, unsigned long addr,
 #endif
 pte_t *find_linux_pte_or_hugepte(pgd_t *pgdir, unsigned long ea,
      unsigned *shift);
+
+static inline pte_t *lookup_linux_ptep(pgd_t *pgdir, unsigned long hva,
+         unsigned long *pte_sizep)
+{
+ pte_t *ptep;
+ unsigned long ps = *pte_sizep;
+ unsigned int shift;
+
+ ptep = find_linux_pte_or_hugepte(pgdir, hva, &shift);
+ if (!ptep)
+  return NULL;
+ if (shift)
+  *pte_sizep = 1ul << shift;
+ else
+  *pte_sizep = PAGE_SIZE;
+
+ if (ps > *pte_sizep)
+  return NULL;
+
+ return ptep;
+}
 #endif /* __ASSEMBLY__ */
 
 #endif /* __KERNEL__ */
-- 
1.7.1